actions/transactWriteItems.js

```diff
@@ -0,0 +1,251 @@
+        }
+
+        // this does NOT ensure atomicity across tables - but the items on each table are already locked
+        // and the actions have been validated. at this point the only thing that would fail would be the
+        // database itself, and that's a lot of work to get around so I'm just ¯\_(ツ)_/¯ not gonna do that
+        var operationsbyTable = Object.entries(batchOpts)
+
+        async.each(operationsbyTable, function([tableName, ops], callback) {
+            var itemDb = store.getItemDb(tableName)
+
+            store.getTable(tableName, function(err, table) {
+                indexUpdates[tableName].forEach(update => {
+                    db.updateIndexes(store, table, update.existingItem, update.item, function(err) {
+                        if (err) return callback(err)
+                    })
+                })
+                itemDb.batch(ops, function(err, results) {
+                    if (err) callback(err)
+                    callback(results)
+                })
+            })
+        }, function(err) {
+            releaseLocks.forEach(release => release()())
+            if (err) return cb(err)
+
+            var res = {UnprocessedItems: {}}, tableUnits = {}
+
+            if (~['TOTAL', 'INDEXES'].indexOf(data.ReturnConsumedCapacity)) {
+                operationsbyTable.forEach(([table, operations]) => {
+                    tableUnits[table] = 0
+                    operations.forEach(op => {
+                        let readCapacity = db.capacityUnits(op.value, true, true)
+                        let writeCapacity = db.capacityUnits(op.value, false, true)
+                        tableUnits[table] += readCapacity + writeCapacity
+                    })
+                })
+                res.ConsumedCapacity = Object.keys(tableUnits).map(function (table) {
+                    return {
+                        CapacityUnits: tableUnits[table],
+                        TableName: table,
+                        Table: data.ReturnConsumedCapacity == 'INDEXES' ? {CapacityUnits: tableUnits[table]} : undefined,
+                    }
+                })
+            }
+
+            cb(null, res)
+        })
+    })
+
+    function addActions(transactItem, cb) {
+        var options = {}
+        var tableName
+
+        if (data.ReturnConsumedCapacity) options.ReturnConsumedCapacity = data.ReturnConsumedCapacity
+
+        if (transactItem.Put) {
+            tableName = transactItem.Put.TableName
+
+            store.getTable(tableName, function (err, table) {
+                if (err) return cb(err)
+                if ((err = db.validateItem(transactItem.Put.Item, table)) != null) return cb(err)
+
+                let value = transactItem.Put.Item
+                let key = db.createKey(transactItem.Put.Item, table)
+                if (seenKeys[key]) {
+                    return cb(db.transactionCancelledException('Transaction cancelled, please refer cancellation reasons for specific reasons'))
```

Comment:
The error text is a little weird...
"refer" -> "refer to" perhaps?
"see reasons for reasons" is kinda circuitous 😆

Reply:
Ugh, that's literally what AWS says 🙈

Reply:
Tell me about it :(